File: ch3/Max1DRangeSum.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
  int n = 9, A[] = { 4, -5, 4, -3, 4, 4, -4, 4, -5 };   // a sample array A
  int running_sum = 0, ans = 0;
  for (int i = 0; i < n; i++)                                       // O(n)
    if (running_sum + A[i] >= 0) {  // the overall running sum is still +ve
      running_sum += A[i];
      ans = max(ans, running_sum);          // keep the largest RSQ overall
    }
    else        // the overall running sum is -ve, we greedily restart here
      running_sum = 0;      // because starting from 0 is better for future
                           // iterations than starting from -ve running sum
  printf("Max 1D Range Sum = %d\n", ans);                    // should be 9
  return 0;
}

File: ch3/Max1DRangeSum.py
n = 9
A = [4, -5, 4, -3, 4, 4, -4, 4, -5]
running_sum = ans = 0
for i in range(n):
    if running_sum + A[i] >= 0:
        running_sum += A[i]
        ans = max(ans, running_sum)
    else:
        running_sum = 0
print("Max 1D Range Sum =", ans)
